### DISCUSSION TITLE
Breaking changes in v0.9.0 changing from `UByteArray` to `String`

### DISCUSSION THREAD

**Discussion by ShreckYe** [ionspin/kotlin-multiplatform-libsodium#40] [Q&A] [answered]
Hi, could you explain the breaking changes in [v0.9.0](https://github.com/ionspin/kotlin-multiplatform-libsodium/releases/tag/0.9.0) that those functions involved return `String` instead of `UByteArray`? From my perspective, hashes should be in binary forms and the old signatures were more natural.

**Comment by ionspin** [comment, marked as answer]
Hi @ShreckYe ,

You are right, it is more natural to work with `UByteArray`, but the underlying [libsodium function crypto_pwhash_str](https://doc.libsodium.org/password_hashing/default_phf#password-storage)  for `PasswordHash.str()` returns a ASCII encoded string and I was previously incorrectly returning a `UByteArray` representation of that string not the real unencoded `UByteArray` of a hash, which was misleading.

**Reply by ShreckYe** [reply]
Thanks for replying. So do you know what kind of encoding it is? Is it Base64 or hex encoding or something else? How do I convert it to and from a `ByteArray`/`UByteArray` if I want the compact form in bytes?

**Reply by ShreckYe** [reply]
OK I see after I printed it. It's a string that also stores the format and parameters, and the hashed part is probably in Base64.

**Reply by ionspin** [reply]
See here for the specification, https://doc.libsodium.org/password_hashing/default_phf#password-storage where it says it's ASCII encoding (https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/to-byte-array.html).
